### PR TITLE
std.logger

### PR DESCRIPTION
**Pull request by burner**
Logging facilities for Phobos.

Docu:
http://burner.github.io/phobos/phobos-prerelease/std_experimental_logger_package.html
http://burner.github.io/phobos/phobos-prerelease/std_experimental_logger_core.html
http://burner.github.io/phobos/phobos-prerelease/std_experimental_logger_filelogger.html
http://burner.github.io/phobos/phobos-prerelease/std_experimental_logger_multilogger.html
http://burner.github.io/phobos/phobos-prerelease/std_experimental_logger_nulllogger.html

dub
http://code.dlang.org/packages/logger

### CODE REVIEW

**Comment by jacob-carlborg** [comment]
The documentation for the alias that uses `defaultLogFunction` still doesn't look good.

File: std/experimental/logger/multilogger.d [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+    {
+        import std.algorithm : copy;
+        import std.range.interfaces : back, popBack;
+        for (size_t i = 0; i < this.logger.length; ++i)
+        {
+            if (this.logger[i].name == toRemove) 
+            {
+                Logger ret = this.logger[i].logger;
+                this.logger[i] = this.logger.back();
+                this.logger.popBack();
```
Comment (burner) [inline std/experimental/logger/multilogger.d]: The sorting of the Logger in the Array does not matter. That this is very fast.

**Comment by burner** [comment]
The dlang doc is changing quite a lot right now. I think I can stay like this for now.

**Comment by jacob-carlborg** [comment]
I don't agree. I think it should be an implementation detail that the logging functions are aliases to `defaultLogFunction`.

**Comment by burner** [comment]
I'm really sure how you want it to change. Could you please create a PR for that.

**Comment by jacob-carlborg** [comment]
To start with, don't use `ditto`.

**Comment by burner** [comment]
thats it?

**Comment by jacob-carlborg** [comment]
It's a start. I assume it's the documentation for the function `defaultLogFunction` you want to share and not the documentation for the template. That's why `ditto` doesn't work so good in this case.

**Comment by burner** [comment]
@jacob-carlborg actually I want to present the trace, info aliases as well because they are supposed to be used. Additionally the template can be used to create new top level log functions/aliases

**Comment by jacob-carlborg** [comment]
@burner I must be really bad at explaining myself. There's a template `defaultLogFunction` containing a function `defaultLogFunction`. The documentation for the template and the function is not the same. When you use `ditto` it will "merge" the documentation for the aliases with the documentation for the template and not the function.

Can you just copy-paste the documentation from the function to the first alias and then use `ditto`?

**Comment by burner** [comment]
@jacob-carlborg I think I get you

**Comment by burner** [comment]
I will have a look at the _loggerconfig.d issue. Thank you

**Comment by burner** [comment]
how should I rebase? One for each commiter?

**Comment by burner** [comment]
@Dicebot I can not reproduce the error. could you give the dmd output

**Comment by burner** [comment]
hm, I build the whole toolchain (my logger branch as phobos) and installed it as my system d compiler. That worked flawlessly with your testcase, with or without the _loggerconfig.d. I used the arch linux aur dmd-git PKGBUILD as a start. I believe the problem is rdmd.

**Comment by burner** [comment]
maybe I was not clear. 

1. I build dmd from dlang/dmd d, druntime from dlang/druntime and phobos from burner/phobos/logger
2. I installed those system wide
3. I created a new folder and placed the test.d file in it. This folder is not a subfolder of phobos or that kind.
4. Than I compiled test.d with the newly installed dmd. (dmd test.d)

actually test_loggerconfig should only be imported if it exists. This is done through a __traits(compiles ) test. That would mean that if it doesn't exists it is not imported. Apparently running dmd directly against a testcase placed in phobos behaves differently.

**Comment by burner** [comment]
dmd: 6b86b12f79e8a0091d641c4e0c294aaae50bb2ab
druntime: feca5225c8bd87a66edaa09db4b2cfc4a655109a
phobos: 299f018

if you use Arch Linux or can use makepkg use this: http://dpaste.dzfl.pl/0d0d8dacac7a and install the packages and continue from step 3. 

again: I doubt that using a non system dmd on a non system phobos will work

**Comment by burner** [comment]
take your time I waited 1.5 Years

**Comment by burner** [comment]
that is very strange. I just copied the test and ran it. it did not import missing.

I tested on two separated machines multiple times in different folders.

**Comment by burner** [comment]
"""
import    foo	(foo.d)
not importing foo
"""

but it does not fail.

**Comment by burner** [comment]
merged burner#4 thanks

**Comment by quickfur** [comment]
Whoa. `std.logger` finally merged?! Time to celebrate!!

**Comment by burner** [comment]
indeed

**Comment by MartinNowak** [comment]
I'm glad to see this in std.experimental.
Can we please get a small announcement in the [changelog](https://github.com/D-Programming-Language/dlang.org/blob/ad64bb378c525402afebe0e0e9f868acc500083e/changelog.dd)?

**Comment by burner** [comment]
@dicebot If you don't mind.
http://forum.dlang.org/post/hmqlrknrjezovnvttxdj@forum.dlang.org should be a good start. thats pretty much what is new.

**Comment by joakim-noah** [comment]
One note: `std.experimental.logger.core` leaves behind a file `someFile.log` when the unittests are run.  I'm not sure what that unit test even does, considering it doesn't contain an assert.

**Comment by burner** [comment]
I will have a look

**Comment by JackStouffer** [comment]
This package has been in `std.experimental` for six releases (3 major, 3 minor) and for seven months. What is the time table for moving it?

**Comment by burner** [comment]
I wish that logger had been moved to into "stable" already, but http://forum.dlang.org/post/mlhj3t$1bdi$1@digitalmars.com sort of blocked it.

**Comment by JackStouffer** [comment]
That's really a shame, it could be a year before a RC *scheme* is decided on let alone it's implementation being added to DMD.

**Comment by burner** [comment]
> it could be a year

I wish I had your optimism. My prediction: RC build into the language will not happen, ever. There will be thread local COW container and the  RefCounted will become better. But there will be no ref counted schema that is build into the compiler.

**Pull request merged**